plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/register-product-collection.tsx

```diff
@@ -0,0 +1,76 @@
+/**
+ * External dependencies
+ */
+import { BlockVariation, registerBlockVariation } from '@wordpress/blocks';
+import { addFilter } from '@wordpress/hooks';
+import { EditorBlock } from '@woocommerce/types';
+import { type ElementType } from '@wordpress/element';
+import type { BlockEditProps } from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import blockJson from '../block.json';
+import {
+	SetPreviewState,
+	PreviewState,
+	ProductCollectionAttributes,
+} from '../types';
+
+export interface ProductCollectionConfig extends BlockVariation {
+	preview?: {
+		setPreviewState?: SetPreviewState;
+		initialPreviewState?: PreviewState;
+	};
+}
+
+/**
+ * Register a new collection for the Product Collection block.
+ *
+ * @param {ProductCollectionConfig} blockVariationArgs The configuration of new collection.
+ */
+const registerProductCollection = ( {
+	preview: { setPreviewState, initialPreviewState } = {},
+	...blockVariationArgs
+}: ProductCollectionConfig ) => {
+	/**
+	 * If setPreviewState or initialPreviewState is provided, inject the setPreviewState & initialPreviewState props.
+	 * This is useful for handling preview mode in the editor.
+	 */
+	if ( setPreviewState || initialPreviewState ) {
+		const withSetPreviewState =
+			< T extends EditorBlock< T > >( BlockEdit: ElementType ) =>
```

Comment:
Is BlockEdit supposed to be the edit component? ElementType would be too broad for that.

It would be more something like ComponentType.

The last defined types in Gberg def typed definitions: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/wordpress__blocks/index.d.ts#L180

(afaict latest blocks doesnt ship with types yet? shrug)

Reply:
I spent some time trying to use `ComponentType<BlockEditProps<T>> | undefined` but it is resulting into more TS errors. I am getting following error, do you have any idea why I am getting this error? 🤔
<img width="873" alt="image" src="https://github.com/woocommerce/woocommerce/assets/16707866/57d9e2a4-cb2f-4d1b-8326-74143581110c">

If it helps, I tried to use `ComponentType` from `'@wordpress/element' as we aren't supposed to import directly from "react", right? 

> (afaict latest blocks doesnt ship with types yet? shrug)

Do you mean, woo blocks doesn't ship with types? 🤷🏻‍♂️

Reply:
> If it helps, I tried to use ComponentType from `'@wordpress/element' as we aren't supposed to import directly from "react", right?

as i mentioned in another comment, you absolutely can import types from react. types are erased at compilation. For some reason wp element is erasing the generics of the ComponentType so the fix here would indeed be to import from react (I tested and it works well).

Reply:
I am getting following error when I try importing from `react`:
![image](https://github.com/woocommerce/woocommerce/assets/16707866/1da1b70c-63ff-4d0c-bc9d-660ae1f2dd05)

I can ignore this lint error by adding following line:
```
// eslint-disable-next-line @typescript-eslint/no-restricted-imports, @woocommerce/dependency-group
```

What do you think? Is it worth it to import "ComponentType" from "react" and then disable eslint for import statement?